### PR TITLE
Implement GET /opportunities/{opportunityId} endpoint

### PR DESCRIPTION
**Pull request by bickelj**
Return a single opportunity when the opportunity id is valid and exists.

Issue #32  Implement GET /opportunities endpoint

### CODE REVIEW

**Comment by bickelj** [comment]
This last push was a correction to the commit comment regarding issue number. It's issue #32 not #31.

**Comment by bickelj** [comment]
Second push had a small stylistic difference in tests to eliminate a duplicate test name.

**Comment by slifty** [comment]
The above push was to rebase onto main

**Review by slifty — CHANGES_REQUESTED** [review]
This is great, and our very first instance of collecting data from the user!

Some thoughts

1. I realized there's a "bug" in our pattern (not a real bug, but a variable naming mistake).
2. I think we would benefit greatly from using ajv and defining request schemas instead of home-brewing validation.

I'll take a swing at implementing these two changes for consideration.

File: src/handlers/opportunitiesHandlers.ts [hunk 1]
```diff
@@ -33,6 +38,50 @@ const getOpportunities = (req: Request, res: Response): void => {
     });
 };
 
+const getOpportunity = (req: Request, res: Response): void => {
+  const isInteger = (it: string): boolean => !Number.isNaN(parseInt(it, 10))
```
[the remark this answers is not in the document]
Comment (slifty) [reply src/handlers/opportunitiesHandlers.ts:42]: I've pushed an implementation as a followup commit -- if it looks good to you I'll squash into the original

File: src/__tests__/opportunities.int.test.ts [hunk 2]
```diff
@@ -59,4 +59,64 @@ describe('/opportunities', () => {
...
+          title,
+          created_at
+        )
+        VALUES
+          ( '🔥', '2525-01-02T00:00:01Z' ),
+          ( '✨', '2525-01-02T00:00:02Z' ),
+          ( '🚀', '2525-01-02T00:00:03Z' )
+        RETURNING id, title;
+      `);
+      logger.debug('opportunityIds: %o', opportunityIds);
```
[the remark this answers is not in the document]
Comment (bickelj) [reply src/__tests__/opportunities.int.test.ts:76]: I wouldn't have thought about not having them until now. What is the reason? Avoiding side effects? Keeping them reasonably clean?

File: src/handlers/opportunitiesHandlers.ts [hunk 3]
```diff
@@ -38,31 +39,44 @@ const getOpportunities = (req: Request, res: Response): void => {
...
+    id: {
+      type: 'integer',
+      minimum: 1,
+    },
+  },
+  required: [
+    'id',
+  ],
+});
+const getOpportunity = (req: Request<GetOpportunityParams>, res: Response): void => {
```
Comment (bickelj) [inline src/handlers/opportunitiesHandlers.ts:57]: While I agree that defining a little `isInteger` function is awkward (there should be an existing easier way), I don't think using AJV to validate a single integer (not JSON) is any less awkward.

**Comment by bickelj** [comment]
> This is great, and our very first instance of collecting data from the user!

Woot!

>     2. I think we would benefit greatly from using ajv and defining request schemas instead of home-brewing validation.

I think we should use Ajv when there is a JSON document to validate, for sure. I am not (yet) sold on the idea of using it to validate a single value is an integer.

File: src/handlers/opportunitiesHandlers.ts (hunk 3 above)
Comment (slifty) [reply src/handlers/opportunitiesHandlers.ts:57]: Worth noting that using a validation library is a common approach to pair with express (and we already have ajv so I didn't wanna lean into joi).  Here's a good writeup if it's useful: https://simonplend.com/how-to-handle-request-validation-in-your-express-api/

The complexity of what we're validating isn't defined by the individual instance but rather the world of possibilities across all endpoints (we want validation errors to endpoint params to be consistent regardless of how complicated the endpoint input is).

Basically though Idon't see ajv as awkward, there are a few functional benefits of using it:

1. consistent error message format across all endpoints
2. ajv is tested and maintained by a large community, so far less risk that we have a bug in our validation / we don't have to test the validators themselves.

File: src/handlers/opportunitiesHandlers.ts [hunk 4]
```diff
@@ -33,6 +38,50 @@ const getOpportunities = (req: Request, res: Response): void => {
...
+      logger.debug(opportunity);
+      if (opportunity.row_count === 0) {
+        res.status(404)
+          .contentType('application/json')
+          .send({ message: 'Not found. Find existing opportunities by calling with no parameters.' });
+        return;
+      }
+
+      const { rows } = opportunity;
+
```
[the remark this answers is not in the document]
Comment (slifty) [reply src/handlers/opportunitiesHandlers.ts:78]: (Just noting for any observers that this discussion is happening more fully over in #74)

File: src/__tests__/opportunities.int.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (slifty) [reply src/__tests__/opportunities.int.test.ts:76]: Yeah I guess cleanliness plus the idea that we don't want to clutter test suite output with logs / logs make it harder to see failing tests.

**Comment by slifty** [comment]
Just to be clear: we aren't using AJV for json anywhere -- the use of "json" in the name AJV is just a reference to the concept of json schema (the mechanism of defining the validation parameters), not the data being validated.

File: src/__tests__/opportunities.int.test.ts [hunk 5]
```diff
@@ -59,4 +59,64 @@ describe('/opportunities', () => {
...
+
+  describe('/id', () => {
+    it('returns exactly one opportunity selected by id', async () => {
+      const opportunityIds: Result<{ id: number; title: string }> = await db.query(`
+        INSERT INTO opportunities (
+          title,
+          created_at
+        )
+        VALUES
+          ( '🔥', '2525-01-02T00:00:01Z' ),
```
[the remark this answers is not in the document]
Comment (bickelj) [reply src/__tests__/opportunities.int.test.ts:71]: It also partially tests that encodings are set properly throughout.

File: src/__tests__/opportunities.int.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bickelj) [reply src/__tests__/opportunities.int.test.ts:76]: Yes, if we need to print statements during tests, we should use the same tools as production.

As for the purpose of printing what came back from the database here, it is to help debug what went wrong when the test fails: was it because the DB sent back an unexpected thing or was it because the subsequent filter or assertion was not set up correctly?

In general, adding these statements and keeping them means you don't have to go and add a log statement or mess around with a debugger, you simply run with debug level and see what happened in more detail. If this is too verbose because the tests run at debug level by default, we could change that to info level.

If these statements in particular are annoying, they can be demoted to trace level or removed.

Does any of that make sense?

Do you usually run with a debugger? Do you usually remove all your log statements before committing?

File: src/handlers/opportunitiesHandlers.ts [hunk 6]
```diff
@@ -33,6 +38,50 @@ const getOpportunities = (req: Request, res: Response): void => {
...
+    && parseInt(it, 10) % 1 === 0;
+
+  if (!isInteger(req.params.id)) {
+    res.status(400)
+      .contentType('application/json')
+      .send({ message: 'Bad request. Send an integer id or call with no parameters.' });
+    return;
+  }
+
+  logger.debug(req);
```
[the remark this answers is not in the document]
Comment (bickelj) [reply src/handlers/opportunitiesHandlers.ts]: I think by default it is logging the fulfilled request, in other words the request and response after it was served. Perhaps that would include enough detail in the error to see what we are trying to see here, perhaps not. The placement of the statement after validation of `req.params.id` is related to its first real use on the following line 53, where it uses the id in the query. If id isn't an integer, it might be unsafe to print the request at all, so that is probably why I had it on the next line. Regardless, I am OK with removing this one and re-adding it locally as needed, or demoting it to trace.

**Comment by bickelj** [comment]
> Just to be clear: we aren't using AJV for json anywhere -- the use of "json" in the name AJV is just a reference to the concept of json schema (the mechanism of defining the validation parameters), not the data being validated.

Not yet, but clearly we should use it to validate requests in JSON where they are. The current use could also be argued to validate a JSO before sending it as JSON proper, as well.

File: src/index.ts [hunk 7]
```diff
@@ -13,6 +13,6 @@ app.listen(
   port,
   host,
   () => {
-    logger.info(`Server running on ${host}:${port}`);
+    logger.info(`Server running on http://${host}:${port}`);
```
Comment (bickelj) [inline src/index.ts:16]: This seems unrelated but I am OK with letting this small change ride along here.

**Comment by bickelj** [comment]
I am not strongly opposed to squashing two authors commits into one but I think it makes the most sense to let the history be the history in most cases, such as this case when there are two authors on one branch. Regardless, I am happy to squash them all into one commit if you would prefer.

File: src/__tests__/opportunities.int.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (slifty) [reply src/__tests__/opportunities.int.test.ts:76]: Totally makes sense

Re do I remove: if they're one-off statements that I used to create the thing the first time or debug an ad hoc issue I remove them.

If they're statements that are fundamental to the continuing evaluation of the health of the project I keep them as you described.

I was mostly asking because it would be good to have a consistent practice around where to add logs when it comes to tests -- there are plenty of tests that make DB calls, so maybe we should make sure to log the result in all cases?  More generally I like the idea of NOT logging manually and just configuring tinypg to do a logger.trace as part of an `onResult` hook (see the [hooks documentation](https://github.com/dynajoe/tinypg#hooks)).

I think logs it's important for logs to be silent during testing in a CI / CD environment (and I personally keep them silent locally) so the test output is unfettered but as you said that's why log levels are configurable!

[the remark this answers is not in the document]
Comment (slifty) [reply src/__tests__/opportunities.int.test.ts:76]: (bottom line this is resolved -- keeping it! -- we might consider that universal change another day though if you like the sound of it)

File: src/__tests__/opportunities.int.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (slifty) [reply src/__tests__/opportunities.int.test.ts:71]: Oh double brilliant

**Comment by slifty** [comment]
I've made the changes -- I did NOT rename that "results" variable here and will just move that patch to the other PR which means we're unblocked on merging this as that question gets resolved.

**Comment by slifty** [comment]
I'm 99% sure that I did NOT change things where there was potential preference to keep the original -- I added the new test and added the new validation mechanism!  With that I'll merge since @bickelj is in new parent mode!

**Review by slifty — APPROVED** [review]
I think this is looking good and that the points of discussion have all been resolved!
